src/main/java/com/ge/predix/uaa/token/lib/ZacTokenService.java

```diff
@@ -47,29 +49,24 @@
 
     @Override
     protected FastTokenServices getOrCreateZoneTokenService(final String zoneId) {
+        return this.tokenServicesMap.get(zoneId);
+    }
+
+    protected FastTokenServices createFastTokenService(final String zoneId) {
         FastTokenServices tokenServices;
-        tokenServices = this.tokenServicesMap.get(zoneId);
         String trustedIssuersURL = this.zacUrl + "/v1/registration/" + getServiceId() + "/" + zoneId;
-        if (null == tokenServices) {
-            try {
-                final ResponseEntity<TrustedIssuers> responseEntity = this.oauth2RestTemplate
-                        .getForEntity(trustedIssuersURL, TrustedIssuers.class);
-                tokenServices = createFastTokenService(responseEntity.getBody().getTrustedIssuerIds());
-                this.tokenServicesMap.put(zoneId, tokenServices);
-            } catch (Exception e) {
-                LOGGER.error("Failed to get trusted issuers from: " + trustedIssuersURL);
-                LOGGER.error(e.getMessage());
-                throw e;
-            }
-
+        try {
+            final ResponseEntity<TrustedIssuers> responseEntity = this.oauth2RestTemplate
+                    .getForEntity(trustedIssuersURL, TrustedIssuers.class);
+            tokenServices = super.createFastTokenService(responseEntity.getBody().getTrustedIssuerIds());
```

Comment:
Does this always return a non null object now?

Reply:
createFastTokenService() will be called by the loading cache if token service for zone is not found in the cache or it expired.  In that case, token service will always be created and inserted in the cache unless exception is thrown.  

According to [this](https://github.com/ben-manes/caffeine/wiki/Population) Note that a cache.get may return null if the entry was not computable or throw an exception if the computation failed.

As discussed, I will add a test to make sure exception is propagated from the cache to the caller of getOrCreateZoneTokenService()

Reply:
@Bitty9, implemented a test for cache exception.  Please review again.